library/ns1_zone.py

```diff
@@ -444,22 +446,49 @@ def get_changed_params(self, have, want):
         :param want: Desired end state of parameters
         :type want: dict
         :return: Parameters in want that differ from have
-        :rtype: dict
+        :rtype: tuple(dict, dict)
         """
         diff = self.diff_params(have, want)
+        result_diff = {}
+
         # perform deep comparison of secondaries if primary exists and has diff
         if "primary" in have and "primary" in diff:
             have_secondaries = have["primary"].get("secondaries")
             want_secondaries = want["primary"].get("secondaries")
+
             # if no difference in values, remove secondaries from diff results
             if not self.diff_in_secondaries(
                 have_secondaries, want_secondaries
             ):
                 diff["primary"].pop("secondaries", None)
+
                 # if secondaries was only key in primary, remove primary
                 if not diff["primary"]:
                     diff.pop("primary", None)
-        return diff
+
+        if self.module._diff:
+
+            # build after dict from have with updated changes
+            after = {}
+            for param in have:
+                if param not in diff:
+                    after[param] = have[param]
+                else:
+                    after[param] = diff[param]
+
+            # convert dictionaries to yaml txt dump
+            have_yaml = yaml.safe_dump(have, default_flow_style=False)
+            diff_yaml = yaml.safe_dump(after, default_flow_style=False)
+
+            # build diff from different yaml dumps
+            result_diff = dict(
+                before=have_yaml,
+                after=diff_yaml
+            )
+            return result_diff, diff
+
+        else:
+            return result_diff, diff
```

Comment:
we could `return {}, diff` here and not need to initialize `result_diff` on :452. Not a really strong opinion but I'd find it a bit clearer. Or moving the variable init a bit closer to where it's used.

Reply:
This is cleaner for sure, making the change.